### PR TITLE
refactor (hasher): update anonymous component feature introduced in circom 2.1.0

### PR DESCRIPTION
**Pull request by C1em3nt**
## Description
This PR mainly focus on changing circuit code style.
 
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally in root directory and any changes were pushed
- [x] Lint (`yarn lint:fix`) was run locally in root directory and any changes were pushed

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [X] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

This PR complete anonymous component change in identity.circom

Previous Version
```
template A(n){
   signal input a, b;
   signal output c;
   c <== a*b;
}
template B(n){
   signal input in[n];
   signal out;
   component temp_a = A(n);
   temp_a.a <== in[0]; 
   temp_a.b <== in[1];
   out <== temp_a.c;
}
component main = B(2);
```

New version
```
template A(n){
   signal input a, b;
   signal output c;
   c <== a*b;
}
template B(n){
   signal input in[n];
   signal out <== A(n)(in[0],in[1]);
}
component main = B(2);
```

By doing so, we could increase the readability of the code and make the code much neater to maintain.

Issue Number: #530 

## What is the new behavior?

The behavior of the modification should not change any result from the previous version.  

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: packages/circuits/circuits/hasher.circom [hunk 1]
```diff
@@ -51,28 +43,18 @@ template StateTreeLeaf(FIELD_COUNT) {
...
-  leaf_identity_hash.inputs[0] <== identity_secret;
-  leaf_identity_hash.inputs[1] <== control;
-
-  component final_hasher = Poseidon(2);
-  final_hasher.inputs[0] <== leaf_identity_hash.out;
-  final_hasher.inputs[1] <== data_hash;
-
-  out <== final_hasher.out;
+  data_hash <== hasher[FIELD_COUNT-2];
+  signal interm <== Poseidon(2)([identity_secret, control]);
```
Comment (moven0831) [inline packages/circuits/circuits/hasher.circom:58]: `leaf_identity_hash` would provide more info about this variable compared to `interm`

File: packages/circuits/circuits/hasher.circom [hunk 2]
```diff
@@ -24,19 +18,17 @@ template EpochTreeLeaf(FIELD_COUNT) {
...
   signal output out;
 
-  component hasher[FIELD_COUNT];
+  signal hasher[FIELD_COUNT];
 
   for (var x = 0; x < FIELD_COUNT; x++) {
-    hasher[x] = Poseidon(2);
     if (x == 0) {
-      hasher[x].inputs[0] <== epoch_key;
+      hasher[0] <== Poseidon(2)([epoch_key, data[0]]);
```
Comment (moven0831) [inline packages/circuits/circuits/hasher.circom:25]: `hasher[x] <== Poseidon(2)([epoch_key, data[0]]);` might provide more consistency to previous work.

File: packages/circuits/circuits/hasher.circom [hunk 3]
```diff
@@ -51,28 +43,18 @@ template StateTreeLeaf(FIELD_COUNT) {
...
   signal output data_hash;
 
-  component hasher[FIELD_COUNT-1];
+  signal hasher[FIELD_COUNT-1];
 
   for (var x = 0; x < FIELD_COUNT-1; x++) {
-    hasher[x] = Poseidon(2);
     if (x == 0) {
-      hasher[x].inputs[0] <== data[0];
+      hasher[0] <== Poseidon(2)([data[0], data[x+1]]);
```
Comment (moven0831) [inline packages/circuits/circuits/hasher.circom:50]: `hasher[x] <== Poseidon(2)([data[0], data[x+1]]);` might provide more consistency to previous work.

**Review by moven0831 — COMMENTED** [review]
One suggestion about the variable name. Two minor suggestions about coding style. Others look good to me, great job!

**Review by FoodChain1028 — APPROVED** [review]
LGTM for now!

**Pull request merged**